BTT SKR V2.0 (STM32F407VGT6) & Stepper Driver Anti-SNAFU Protection

thisiskeithb opened the pull request:
### Description

<img src="https://user-images.githubusercontent.com/13375512/114345141-07f4c680-9b16-11eb-895d-df6c2da52e63.png" width="30%" align="right">

The SKR V2 is the next board in the SKR V1.1/1.3/1.4 lineup, but is now fitted with a 168MHz STM32F407VGT6 and some of the following upgrades:

- Driver ~~anti-reverse~~ Anti-SNAFU Protection
- Thermistor protection circuit (as found in the SKR E3 Turbo)
- WiFi (with an ESP addon)
- USB flash drive support
- **TMC DIAG jumpers are back**! You no longer have to cut or desolder driver diag pins if you're running endstops instead of sensorless homing.

#### To-Do

- [x] Use generic variant instead of duplicating more files into Marlin
- [x] Possibly rename "Driver Anti-Reverse Protection" feature for simplicity?
- [x] Translatable error message

### Requirements

SKR V2

### Benefits

SKR V2 board support & driver protection.

### Configurations

[SKR_2_Test_Config.zip](https://github.com/MarlinFirmware/Marlin/files/6328913/SKR_2_Test_Config.zip)

### Related Issues

n/a

thisiskeithb committed:
Driver Anti-Reverse Protection

thisiskeithb committed:
SKR V2.0 Support

thisiskeithb commented:
The “Driver Anti-Reverse Protection” name was brought over from BigTreeTech’s original code, but maybe that can be shortened to “Driver Protection” instead?

rhapsodyv commented:
Marlin already have a generic F4x7 variant. Is there any reason to add one specific for this board?

thisiskeithb commented:
> Marlin already have a generic F4x7 variant. Is there any reason to add one specific for this board?

It’s what BigTreeTech did 🤷‍♂️

I think you and I tested the generic variant on this board a while ago and it didn’t work, but we can do more testing on discord if you want to remove the new variant.

rhapsodyv commented:
> > Marlin already have a generic F4x7 variant. Is there any reason to add one specific for this board?
> 
> It’s what BigTreeTech did 🤷‍♂️
> 
> I think you and I tested the generic variant on this board a while ago and it didn’t work, but we can do more testing on discord if you want to remove the new variant.

It's good for our sanity if we keep custom variants at minimum. 

Yeah, lets do more tests. Probably was just a missing parameter.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

Msq001 committed:
Mass production of pin. Avoid sharing TMC SPI with SD SPI

Msq001 committed:
Enable anti-reverse protection in default if SAFE_POWER_PIN is defined

bigtreetech commented:
Hello, I made some small changes in https://github.com/thisiskeithb/Marlin/pull/5, please check.

thisiskeithb committed:
Merge pull request #5 from bigtreetech/pr/btt_skr_v2_pin

Pr/btt skr v2 pin

thisiskeithb committed:
Fix merge

thisiskeithb committed:
Temporarily add generic SKR V2 variant

For use while debugging

thisiskeithb commented:
My last commit temporarily adds the generic SKR V2 variants to make it easier for other V2 beta testers to test this PR.

Once rhapsodyv and I work through the remaining issues, the custom variant & associated files will be removed.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Update generic SKR V2 usb env

thisiskeithb committed:
Remove generic SKR V2 env

Remove for now. Need to spend more time tracking down timer/pins/display issues.

thisiskeithb commented:
@rhapsodyv and I can't get this board working correctly in the generic environment, so I've removed that for now.

Other than my question about possibly renaming the driver anti-reversal feature, possibly adding the ability for the error messages to be translated/wording cleanup, I think this PR is nearly ready.

thisiskeithb commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -1628,6 +1628,13 @@ static_assert(hbm[Z_AXIS] >= 0, "HOMING_BUMP_MM.Z must be greater than or equal
+/**
+ * Stepper Driver Anti-Reverse Protection
+ */
+#if ENABLED(DRIVER_SAFE_POWER_PROTECT) &&  !PIN_EXISTS(SAFE_POWER)
```
I don't think `PIN_EXISTS(SAFE_POWER)` is ever evaluated here. I can comment out the `SAFE_POWER_PIN` and no error is thrown.

bigtreetech replied:
Yes, nothing will be enabled without `SAFE_POWER_PIN` definition, so this is redundant, and I deleted it in https://github.com/thisiskeithb/Marlin/pull/6

thisiskeithb commented on Marlin/src/inc/Conditionals_adv.h:
```diff
@@ -543,3 +543,7 @@
   #define HAS_MEATPACK 1
 #endif
+
+#if DISABLED(DISABLE_DRIVER_SAFE_POWER_PROTECT) && PIN_EXISTS(SAFE_POWER)
```
I don't think `PIN_EXISTS(SAFE_POWER)` is ever evaluated here either since a single `#if PIN_EXISTS(SAFE_POWER) #error "safe power pin found"` check doesn't throw an error.

bigtreetech replied:
It seems that the `PIN_EXISTS` macro can't be called in `Conditionals_adv.h`.  So I moved this judgment to `Conditionals_post.h` file in https://github.com/thisiskeithb/Marlin/pull/6

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

rhapsodyv committed:
fix flash drive env

thinkyhead committed:
Check pin_exists after pins include

Msq001 committed:
fix SanityCheck

thisiskeithb committed:
Merge pull request #6 from bigtreetech/pr/btt_skr_v2

fix SanityCheck.h

thinkyhead committed:
adjustments

thinkyhead commented:
The term "reverse" in relation to drivers implies something related to the DIR pin. So, I have substituted "backwards" for "reverse," in the process dropping all noun-ish instances of "anti." Other terms like "rotated," "flipped," and "twisted" may still be considered.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

Msq001 committed:
variants

bigtreetech commented:
@thisiskeithb @rhapsodyv 
Hello, I don't yet know why you canceled the decision to use the generic `MARLIN_F4x7Vx` variant of SKR-2
 I submitted new changes in https://github.com/thisiskeithb/Marlin/pull/7, and the content I tested, please review

thisiskeithb committed:
Merge pull request #7 from bigtreetech/pr/btt_skr_v2

SKR-2 use Generic variant

bigtreetech commented:
I forgot to delete `BIGTREE_SKR_2` variant in `\buildroot\share\PlatformIO\variants`, please delete it. Thank you bery much.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb committed:
Remove SKR_2 variant

thisiskeithb commented:
> Hello, I don't yet know why you canceled the decision to use the generic MARLIN_F4x7Vx variant of SKR-2

@rhapsodyv and I were trying to convert over to the generic variant, but it was difficult due to the remote nature of the work. I'll get my board flashed and confirm everything works.

> I forgot to delete BIGTREE_SKR_2 variant in \buildroot\share\PlatformIO\variants, please delete it. Thank you bery much.

Removed in https://github.com/MarlinFirmware/Marlin/pull/21503/commits/cb7d2c3cfecbac16f1896eb3347bdba06466a208

bigtreetech commented on buildroot/share/PlatformIO/boards/BigTree_SKR_2.json:
```diff
@@ -0,0 +1,46 @@
+{
+  "build": {
```
This board Json file is also no longer needed.

thisiskeithb replied:
https://github.com/MarlinFirmware/Marlin/pull/21503/commits/460e97748980a6c79e282424e7636f6bb9bdfaf4

bigtreetech commented on Marlin/src/core/boards.h:
```diff
@@ -363,25 +363,27 @@
-#define BOARD_BTT_SKR_PRO_V1_1        4207  // BigTreeTech SKR Pro v1.1 (STM32F407ZG)
-#define BOARD_BTT_SKR_PRO_V1_2        4208  // BigTreeTech SKR Pro v1.2 (STM32F407ZG)
-#define BOARD_BTT_BTT002_V1_0         4209  // BigTreeTech BTT002 v1.0 (STM32F407VG)
+#define BOARD_BTT_SKR_PRO_V1_1        4207  // BigTreeTech SKR Pro v1.1 (STM32F407VGT6)
```
The MCU of SKR Pro V1.1 & 1.2 is `STM32F407ZGT6`

thisiskeithb replied:
https://github.com/MarlinFirmware/Marlin/pull/21503/commits/460e97748980a6c79e282424e7636f6bb9bdfaf4

thisiskeithb committed:
Some cleanup

thisiskeithb commented:
It could be because I'm using a beta SKR V2 board, but my non-emulated `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER` display is garbled and I am getting TMC CONNECTION errors with the generic variant. These are the same issues that rhapsodyv and I were running into.

If I revert back to the original, custom variant, then everything works as it should.

thisiskeithb committed:
Translatable on-screen error message

gordo3di commented:
@thisiskeithb @bigtreetech 

I just got my retail skr v2's in yesterday and used your repository. Everything seems to work great and I'm just trying to figure our the EEPROM emulation. Otherwise thank you and great work! Do you know when this additions might be pulled into Marlin officially?

thisiskeithb commented:
> Everything seems to work great and I'm just trying to figure out the EEPROM emulation. Otherwise thank you and great work! Do you know when this additions might be pulled into Marlin officially?

It's already supported. Enable `EEPROM_SETTINGS` in your config.

calderon67 commented:
> > Everything seems to work great and I'm just trying to figure out the EEPROM emulation. Otherwise thank you and great work! Do you know when this additions might be pulled into Marlin officially?
> 
> It's already supported. Enable `EEPROM_SETTINGS` in your config.

I'm sorry, but I must say that the firmware keeps on not working. What I have a prusa i3 steel with nema17 and TMC2130. On initial state, only moving motors without heating anything, the m122 debug info is normal and printer works normally. However, when I try to do autolevel with a bltouch after heating and maintaining bed at 70ºC, X motor stop moving, and m122 report of X driver seems to be not normal. It keeps on with the same issue reported before.

Remember that before I changed wire, driver and motor with the same result.

thisiskeithb commented:
> I'm sorry, but I must say that the firmware keeps on not working.

I don't have a retail SKR V2 to troubleshoot with (and won't for a couple more weeks), so I'll have to defer to @bigtreetech.

gordo3di commented:
> > I'm sorry, but I must say that the firmware keeps on not working.
> 
> I don't have a retail SKR V2 to troubleshoot with (and won't for a couple more weeks), so I'll have to defer to @bigtreetech.

Where are you located? We can send you a board.

calderon67 commented:
I can confirm that it seems to print normally without using the heated bed

rhapsodyv commented:
We could have an issue in STM32 HAL with SPI drivers. I don't have any SPI driver here to test. I will try to get some and check if the issue is with stm32 hal.

gordo3di commented:
@thisiskeithb 
What are you putting here? If I disable both then it says no EEPROM when I save bed leveling...

```
#if NO_EEPROM_SELECTED
  //#define SRAM_EEPROM_EMULATION                 // Use BackSRAM-based EEPROM emulation
  //#define FLASH_EEPROM_EMULATION                  // Use Flash-based EEPROM emulation
#endif
```

thisiskeithb commented:
> What are you putting here?

I used the provided pins file which has: 

```cpp
// Use one of these or SDCard-based Emulation will be used
#if NO_EEPROM_SELECTED
  //#define SRAM_EEPROM_EMULATION                 // Use BackSRAM-based EEPROM emulation
  #define FLASH_EEPROM_EMULATION                  // Use Flash-based EEPROM emulation
#endif
```

@bigtreetech Does emulated EEPROM work in your tests?

bigtreetech commented:
> @bigtreetech Does emulated EEPROM work in your tests?

`FLASH_EEPROM_EMULATION` is normal in my basic test.

thisiskeithb commented:
@gordo3di with your other reported issues, your board may have something else wrong with it.

gordo3di commented:
> @gordo3di with your other reported issues, your board may have something else wrong with it.
@thisiskeithb 
I have 5 more boards so today I'll be installing a few and doing lots more testing.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

Msq001 committed:
delete onboard sd detect pin

thisiskeithb committed:
Remove onboard sd detect pin

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

rhapsodyv commented on Marlin/src/MarlinCore.cpp:
```diff
@@ -1220,6 +1224,15 @@ void setup() {
   #endif
 
+  #if PIN_EXISTS(SAFE_POWER)
+    #if ENABLED(HAS_DRIVER_SAFE_POWER_PROTECT)
```
I think that when we use HAS_SOMETHING, we don't use ENABLED. The HAS_SOMETHING usually has a integer value 1 to work direct with #if

thisiskeithb committed:
no enabled()

rhapsodyv commented:
I think we can merge this, to allow more users to test.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb commented:
> fatal error: LiquidCrystal.h: No such file or directory

I don't know why it's complaining about a `LiquidCrystal` dependency all of a sudden...

rhapsodyv commented:
Maybe a issue with platformio-dev? Try change CI to use normal pio and check if it works.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

thisiskeithb commented:
> Maybe a issue with platformio-dev? Try change CI to use normal pio and check if it works.

Wasn't me! 🙂 https://github.com/MarlinFirmware/Marlin/commit/993f9f3efd4e4d4feeeb676f110de2ca72a20622

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

calderon67 commented:
Has been released any patch update for SPI TMC motor driver problem?

rhapsodyv commented:
> Has been released any patch update for SPI TMC motor driver problem?

No, I don't have any spi tmc driver yet. It will take a while until I get some.

calderon67 commented:
> > Has been released any patch update for SPI TMC motor driver problem?
> 
> No, I don't have any spi tmc driver yet. It will take a while until I get some.

Ok, don't worry. Please, post here anyone any update related to that. 

Thank you all very much for your contribution!

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/btt_skr_v2

gordo3di commented:
> > Has been released any patch update for SPI TMC motor driver problem?
> 
> No, I don't have any spi tmc driver yet. It will take a while until I get some.

Sorry. What was the SPI TMC issue? I've been running TMC2130's on several printers with the v2 board without issue... Flash eeprom also works fine now. Honestly the v2 boards have been running great so far...

calderon67 commented:
> > > Has been released any patch update for SPI TMC motor driver problem?
> > 
> > 
> > No, I don't have any spi tmc driver yet. It will take a while until I get some.
> 
> Sorry. What was the SPI TMC issue? I've been running TMC2130's on several printers with the v2 board without issue... Flash eeprom also works fine now. Honestly the v2 boards have been running great so far...

On my case it doesn't work. Several axes stop working in different moments. At the very beginning it started with the X axes and autolevel feature. When autolevel was commanded without maintaining any temperature on heated bed everything seems to work normally. However, when a bed temperature was commanded, after reaching being maintaining it, suddenly X axes stopped during autolevel.

I also tested to print something (a 10mm cube) without using bed temperature, but suddenly Z axis stop working just like X.

I have commanded M122 for debugging and this was the results:

![fault](https://user-images.githubusercontent.com/16945237/116261278-c69c2200-a777-11eb-8d9a-6439c91660c7.png)

I have tried everything. I have tried changing stepper driver, wire, and also stepper motor, but X axes failure during autolevel with heated bed at 70ºC keeps on. I also have no problem on sending configuration.h and configuration.adv if you find it interesting for solving the problem.

After resetting the board everything seems to be once again working well, and this is the result for M122:

![normal](https://user-images.githubusercontent.com/16945237/116261105-9e142800-a777-11eb-8bbf-fa2352f650c8.png)


By the moment I do not understand what's going on, but it doesn't seem to me that it is related with hardware.

thisiskeithb commented:
@calderon67: Reach out to BigTreeTech's support to make sure it's not a hardware issue. If they deem the board fine, feel free to come back and post a [Bug Report](https://github.com/MarlinFirmware/Marlin/issues/new/choose).

calderon67 commented:
> @calderon67: Reach out to BigTreeTech's support to make sure it's not a hardware issue. If they deem the board fine, feel free to come back and post a [Bug Report](https://github.com/MarlinFirmware/Marlin/issues/new/choose).

I have written an email to them, but I am pretty sure that it is not a hardware problem. The main reason is that, after getting X failure and when printer thinks that autolevel has been finished, it does not respond to any movement command ordered via Pc or via TFT35. After that, without touching or moving anything, mainly the main board, I press reset button on TFT and after that X axes moves normally once again.

If you don't mind I would prefer to open a bug report at the same time.

Thank you all very much

gordo3di commented:
@calderon67 

Your current is much lower than mine. We make very large machines (18" bed size) so I set our current to 1000ma but this means we absolutely need good cooling on the drivers. Before I added the cooling (in this case a blower fan) the drivers would drop current incrementally causing layer shifting and eventually stopping the drivers. Maybe this is the issue?

I used @thisiskeithb 's fork from a last week and pretty much only changed the configs for our machines.

calderon67 commented:
> @calderon67
> 
> Your current is much lower than mine. We make very large machines (18" bed size) so I set our current to 1000ma but this means we absolutely need good cooling on the drivers. Before I added the cooling (in this case a blower fan) the drivers would drop current incrementally causing layer shifting and eventually stopping the drivers. Maybe this is the issue?
> 
> I used @thisiskeithb 's fork from a last week and pretty much only changed the configs for our machines.

My printer was previously running with ramps with this TMC on SPI, and I used 350ma for X and Y, 700ma for the driver of the two Z stepper motors and 800ma for E. It worked normally with that current set up. My printer is a Prusa i3 with a heated bed of 200mm * 200mm.

Anyway, I have added a 50mm fun just in front of the drivers, and the X driver is the first one on air blow line. Also it should be taken into account that overheat make no sense because it doesn't trigger overheat prewarn tag of TMC2130 and because autolevel is mostly the first moving command, which is not very intense. We are talking about autolevel grid of 9 points.

I do not think it is related with that, but I have not tested to increase ma of the driver. However, why does it suddenly change msteps from 16 to 256? And why does it suddenly change X driver enabled to false?

It makes no sense to me

thisiskeithb commented:
@calderon67 & @gordo3di: Since this PR has been merged & it's not an obvious issue for *all* SKR V2 boards, it'd be best to take this discussion to the [Marlin Forum](https://forums.reprap.org/list.php?415) hosted on RepRap.org or the [Marlin Firmware](https://discord.gg/n5NJ59y) Discord server.

Guilouz commented:
Hi, I just received this from BigTreeTech. Does it means this PR not working correctly ?

Dear friend, I am sorry for the inconvenience. We found that a considerable number of users have inserted the stepper driver in the wrong direction. This may cause their stepper driver and motherboard to be damaged.
In response to this problem, we provide the following three solutions for you to choose:

Option 1: On May 10th, we will arrange the latest repaired and upgraded SKR 2 for you

Option 2:
We will arrange to send a new MOSFET to you and help to guide
you through the replacement process which will require soldering
skills. we would like to offer a discount of 10% on any over $20 product from BIGTREETECH. And we will send the free ESP8266 WiFi module that can be used with the SKR 2.
SKR 2 when running Marlin and RRF (the module will be sent with
the MOSFET). (Valid discount time is 1 year and the SKR 2
warranty will be extended to 1 year from the date)

Option 3:
Choose to fix it quickly:
If you don’t want to wait for the new MOSFET and need to use the
SKR 2 urgently then we have created a guide which shows
you how to safely use it with TMC drivers. If you are prepared to
follow the guide, then we will offer you a 50% refund on your
board by paypal
We will reply to you on May 4th, because May 1-3 is our Chinese holiday. Sorry again, thank you for your understanding

thisiskeithb commented:
> @thisiskeithb Hi, I just received this from BigTreeTech. Does it means this PR not working correctly ?

No. Contact BigTreeTech for a replacement if you're having issues.

thisiskeithb commented:
> I just received this message from BTT Store following my purchase, that's why I was wondering.

Yes, please do as the email instructed and reach out the BigTreeTech as it's not related to Marlin.
